pinot-core/src/main/java/org/apache/pinot/core/transport/DirectOOMHandler.java

```diff
@@ -0,0 +1,100 @@
+// no one will reach channelRead0.
+public class DirectOOMHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataTableHandler.class);
+  private static final AtomicBoolean DIRECT_OOM_SHUTTING_DOWN = new AtomicBoolean(false);
+  private final QueryRouter _queryRouter;
+  private final ServerRoutingInstance _serverRoutingInstance;
+  private final ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> _serverToChannelMap;
+
+  public void setSilentShutDown() {
+    _silentShutDown = true;
+  }
+
+  private volatile boolean _silentShutDown = false;
+
+  public DirectOOMHandler(QueryRouter queryRouter, ServerRoutingInstance serverRoutingInstance,
+      ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> serverToChannelMap) {
+    _queryRouter = queryRouter;
+    _serverRoutingInstance = serverRoutingInstance;
+    _serverToChannelMap = serverToChannelMap;
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    // if we are shutting down channels due to direct memory OOM, we short circuit the channel inactive
+    if (_silentShutDown) {
+      return;
+    }
+    ctx.fireChannelInactive();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+    // catch direct memory oom here
+    if (cause instanceof OutOfMemoryError && cause.getMessage().contains("Direct buffer")) {
+      // only one thread can get here and do the shutdown
+      if (DIRECT_OOM_SHUTTING_DOWN.compareAndSet(false, true)) {
+        try {
+          LOGGER.error("Closing ALL channels to servers, as we are running out of direct memory "
+              + "while receiving response from {}", _serverRoutingInstance, cause);
+          // close all channels to servers
+          ctx.channel().close();
+          _serverToChannelMap.keySet().forEach(serverRoutingInstance -> {
+            Channel channel = _serverToChannelMap.get(serverRoutingInstance)._channel;
+            if (channel != null) {
+              DirectOOMHandler directOOMHandler = channel.pipeline().get(DirectOOMHandler.class);
+              if (directOOMHandler != null) {
+                directOOMHandler.setSilentShutDown();
+              }
+              channel.close();
+            }
+            _serverToChannelMap.remove(serverRoutingInstance);
+          });
+          _queryRouter.markServerDown(_serverRoutingInstance,
+              new QueryCancelledException("Query cancelled as broker is out of direct memory"));
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while handling direct memory OOM", e);
+        } finally {
+          DIRECT_OOM_SHUTTING_DOWN.set(false);
+        }
+      } else {
+        LOGGER.warn("Caught direct memory OOM, but another thread is already handling it", cause);
+      }
+    } else {
+      ctx.fireExceptionCaught(cause);
```

Comment:
the default implementation of this method is to do `ctx.fireExceptionCaught(cause)`, which basically notifies the next handler in the pipeline. So in case we override the method, we need to do this call. Think about that like the try-catch-throw pattern in normal Java code. If your handler deals with the exception, then is fine to do not _throw_ (aka `fireExceptionCaught`)

Reply:
Got it. Thanks